packages/discovery-react-components/src/utils/document/documentUtils.ts

```diff
@@ -208,6 +204,86 @@ export function uniqRects(rects: DOMRectList): Partial<DOMRectList> {
   );
 }
 
+/**
+ * Finds the descendant node that contains the given offset.
+ * This is done efficiently via a binary search.
+ * @param parentNode node to search the descendants of
+ * @param offset value to find within the nodes
+ * @return the descendant node, or null if none is found
+ */
+function findContainingNodeWithin(parentNode: HTMLElement, offset: number) {
+  let foundNode = null,
+    nodeToSearch: HTMLElement | null = parentNode;
+
+  while (!!nodeToSearch) {
+    // Get all children of the current node to search
+    let nodes = Array.from(nodeToSearch.children) as HTMLElement[];
+
+    // We can assume that all top level nodes have begin and end set
+    // (This is a performance optimization-- can be removed if it causes issues.)
+    if (nodeToSearch === parentNode) {
+      // Make sure that all nodes have begin / end values
+      nodes = nodes.filter(
+        node => node.dataset.childBegin !== undefined && node.dataset.childEnd !== undefined
+      );
+    }
+
+    // Perform binary search for the offset value on the array of nodes
+    // (This array of nodes will always be sorted by both
+    //  begin AND end, since sibling nodes are disjoint.)
+    const foundIndex = indexOfContainingNode(nodes, offset);
```

Comment:
Can your custom binary-search code be replaced with one of the Lodash methods? https://lodash.com/docs/4.17.15#sortedIndex

Reply:
If keeping your custom code, just have it return the node itself or null, instead of just the index.